unidist/core/backends/mpi/core/controller/common.py

```diff
@@ -255,3 +256,76 @@ def push_data(dest_rank, value):
             _push_data_owner(dest_rank, value)
         else:
             raise ValueError("Unknown DataID!")
+
+
+def push_data_owners(dest_rank, data_ids):
+    """
+    Parse and send data owners for all values to destination rank.
+
+    Send all ID associated location to the target rank.
+
+    Parameters
+    ----------
+    dest_rank : int
+        Rank where task data is needed.
+    data_ids : list
+        List of all data_ids required by the worker
+    """
+    for data_id in data_ids:
+        if object_store.contains_data_owner(data_id):
+            _push_data_owner(dest_rank, data_id)
+        else:
+            raise ValueError("Data owner not known")
+
+
+def choose_destination_rank(data_ids):
+    """
+    Choose destination rank considering which worker has the maximum share of data.
+
+    Get the data shares in each worker process and choose rank with the most data share.
+    If data_ids are empty choses the rank using the round robin approach.
+
+    Parameters
+    ----------
+    data_ids : list
+        List of all data_ids required by the worker
+
+    Returns
+    -------
+    int
+        A rank number.
+    """
+    if not data_ids:
+        chosen_rank = RoundRobin.get_instance().schedule_rank()
+        return chosen_rank
+    data_share = defaultdict(lambda: 0)
+    for data_id in data_ids:
+        data_share[object_store.get_data_owner(data_id)] += object_store.get_data_size(
+            data_id
+        )
+    chosen_rank = max(data_share, key=data_share.get)
+    return chosen_rank
+
+
+def collect_all_data_id_from_args(value, collected_data_ids=[]):
+    """
+    Collect all data ids
+
+    Collect all data ids from the given value and save them to  collected_data_ids list.
+
+    Parameters
+    ----------
+    value : iterable or dict or object
+        Arguments to be sent.
+    collected_data_ids : list to store all collected data_ids
+
+    """
+    if isinstance(value, (list, tuple)):
+        for v in value:
+            collect_all_data_id_from_args(v, collected_data_ids)
+    elif isinstance(value, dict):
+        for v in value.values():
+            collect_all_data_id_from_args(v, collected_data_ids)
+    elif is_data_id(value):
```

Comment:
Is it possible if we don't get into any of if branches? You can add raise an exception at the end of the function to check this when testing with Modin.

Reply:
Here as the plan is to just collect data_ids, recursively so if the args are not data_ids, eg functions, python objects etc it is expected not to get into any if branches. Will add an exception if object store does not contain an owner.